python/README.md

```diff
@@ -0,0 +1,139 @@
+# Python in the code-dot-org monorepo
+
+While our monorepo is predominantly geared toward Ruby on Rails, we have the ability to
+run Python code from Rails. This is enabled by [pycall.rb](https://github.com/mrkn/pycall.rb),
+and allows us to take advantage of Python's great package ecosystem. More information is available
+in [the initial pycall PR](https://github.com/code-dot-org/code-dot-org/pull/60048).
+
+For package management and virtual env we use [pdm](https://pdm-project.org/), see
+[section on using pdm below](#pdm-manage-python-packages-and-virtualenv).
+
+## pdm: manage python packages and virtualenv
+
+[pdm](https://pdm-project.org/) uses [pyproject.toml](../pyproject.toml) to create a python virtualenv, and install and manage its
+dependencies. Its similar to `bundle` from the Ruby world, or `yarn`/`npm` from the Node world,
+and includes many of the same features.
+
+Common commands:
+
+- `pdm install`: install dependencies specified in pyproject.toml. like: `yarn install` or `bundle install`
+- `pdm add boto9000`: add boto9000 to pyproject.toml and install it. like: `yarn install boto9000`
+- `pdm run ____`: run `____` inside the repo's python virtualenv. like: `bundle exec ____`
+  - example: `pdm run ipython`: start ipython
+  - example: `pdm run pytest`: run pytest in the current dir
+- `source .venv/bin/activate`: activate the python virtual environment created by pdm (see `pdm venv activate` for exact command)
+
+## pycall.rb: how to invoke python code from rails
+
+- Ruby => Python function calls are very fast: microseconds not milliseconds
+- Ruby callbacks can be passed into Python functions
+- Python packages, specified in sub-packages like `/python/pycdo/pyproject.toml`, can be imported in Ruby using `pyimport('pipname')`
+- Python code can be added to pycdo, e.g. /python/pycdo/mynewfeature.py, or by creating a new toplevel package like /python/myfeature. See [how to structure your python code](#how-to-structure-your-python-code).
```

Comment:
I also added a comment to `locals.yml.default` about switching the queue_adapter to delayed_job:
```
# Uncomment to use `delayed_job` to fulfill ActiveJob requests. This is useful
# for matching prod, and is also currently required to use ActiveJobs that
# invoke Python via PyCall. These will SEGV if run with the default async
# adapter, because it uses threads instead of processes.
#
# If you enable `delayed_job`, you will need to manually run `bundle exec bin/delayed_job start`
# from the dashboard/ directory to start the workers.
#
# active_job: { queue_adapter: delayed_job }
```

Reply:
Oof, a new rabbit hole, have to make that actually configurable via config.yml.erb.

Reply:
Oof indeed! for that rabbit hole, you should be able to set a CDO variable in this file and then pass it here: https://github.com/code-dot-org/code-dot-org/blob/90410de82aae9fe48f0165acbc7d5dcf87b1e205/dashboard/config/application.rb#L223

if pycall works only on the main thread, then one idea would be to wrap our access to it (calls to `pyimport`,  etc) with a ruby check for `Thread.current == Thread.main`, followed by a helpful error message explaining the requirements for running pycall rather than an opaque seg fault.

I don't want to go too crazy with protections given our intent to solve the threading issue, so feel free to put this off though.

Reply:
or perhaps lib/cdo/pycall.rb could simply provide a method
```
def safe_to_run_pycall?
  Thread.main == Thread.current
end
```
which developers can then use in their active jobs before calling into pycall.